# Parte 1: Explicar e explorar

Para cada bloco de código desta aula:

1.   **Explicar:** escreva, com suas palavras, o que cada linha faz e por que ela existe.
2.   **Explorar:** rode o código, altere algo (um documento, um termo, um argumento) e
registre o que mudou no resultado.
3.   **Prever:** antes de rodar a versão alterada, escreva o que você espera que aconteça, e compare.



## Bloco 1: Criação de Vetor

In [ ]:
docs <- c(
  d1 = "recuperacao de informacao ordena documentos por relevancia",
  d2 = "o modelo de espaco vetorial representa documentos como vetores",
  d3 = "bm25 e um modelo probabilistico de ranqueamento de texto",
  d4 = "aprendizado estatistico fundamenta a recuperacao moderna",
  d5 = "o indice invertido acelera a busca em muitos documentos",
  d6 = "embeddings capturam a semantica de palavras e documentos",
  d7 = "a avaliacao mede a relevancia dos resultados da busca",
  d8 = "ciencia de dados combina estatistica e programacao"
)
length(docs)
docs["d5"]

[1] 8

d5 
"o indice invertido acelera a busca em muitos documentos"

### 1. O que cada linha faz:
- docs <- c (...): atribui à variável docs um *combine/concatanate* (c) das 8 frases, cada uma com seu índice (d1 a d8).
- length(docs): exibe a extensão do vetor docs
- docs ["d5"]: exibe a frase gravada no índice d5 do vetor/variável docs.

### 2. Prever e Explorar

- Adicionar um novo elemento

**Previsão:** Ao adicionar d9 na lista, o comando length(docs) deve passar de 8 para 9, e a busca por docs["d9"] retornará a nova frase.

In [ ]:
docs <- c(
  docs,
  d9 = "inteligencia artificial em ciencia de dados"
)
length(docs)
docs["d9"]

[1] 9

d9 
"inteligencia artificial em ciencia de dados"

**Resultado:** length(docs) retornou 9. O vetor expandiu mantendo as chaves anteriores intactas.

## Bloco 2: Tokenização

In [ ]:
tokenizar <- function(texto) {
  texto <- tolower(texto)
  unlist(strsplit(texto, "\\s+"))
}

tokens <- lapply(docs, tokenizar)
tokens[["d1"]]

[1] "recuperacao" "de"          "informacao"  "ordena"      "documentos" 
[6] "por"         "relevancia"

### 1. O que cada linha faz:
- tokenizar <- function(texto): atribui à variável tokenizar uma função que recebe um parâmetro texto.
- texto <- tolower(texto): os caracteres do texto são todos transformados em minúsculas
- unlist(strsplit(texto, "\\s+")): divide em tokens. O \\s+ é regex para um ou mais espaços em branco seguido. O texto é quebrado pelo strsplit onde houver espaços. O unlist "achata" a lista, transformando a resposta em um vetor simples de texto (em vez de uma lista)

### 2. Prever e Explorar

- Alterar o regex para dividir por letras em vez de palavras

**Previsão:** Ao trocar o padrão "\\s+" por "" (uma string vazia) dentro do *strsplit*, o R deixará de separar o texto por espaços e passará a quebrar a frase caractere por caractere, gerando um vetor com todas as letras e espaços individuais..

In [ ]:
tokenizar_letras <- function(texto) {
  texto <- tolower(texto)
  unlist(strsplit(texto, ""))
}

tokens_letras <- lapply(docs, tokenizar_letras)
tokens_letras[["d1"]]

[1] "r" "e" "c" "u" "p" "e" "r" "a" "c" "a" "o" " " "d" "e" " " "i" "n" "f" "o"
[20] "r" "m" "a" "c" "a" "o" " " "o" "r" "d" "e" "n" "a" " " "d" "o" "c" "u" "m"
[39] "e" "n" "t" "o" "s" " " "p" "o" "r" " " "r" "e" "l" "e" "v" "a" "n" "c" "i"
[58] "a"

**Resultado:** A saída de *tokens_letras[["d1"]]* deixou de ser um vetor de 7 palavras e virou um vetor longo com cada letra e espaço individual ("r", "e", "c", "u", "p", "e", "r", "a", "c", "a", "o", " ", ...).

## Bloco 3: Vocabulário e frequência de termos

In [ ]:
vocab <- sort(unique(unlist(tokens)))
length(vocab)

# frequencia total de cada termo no corpus
freq <- table(unlist(tokens))
sort(freq, decreasing = TRUE)[1:6]

[1] 47


        de          a documentos          e      busca    ciencia 
         7          5          4          3          2          2 

### 1. O que cada trecho faz:

vocab <- sort(unique(unlist(tokens)))

- unlist(tokens): Desfaz a estrutura de lista e junta as palavras de todos os documentos em um único vetor gigante.
- unique(...): Remove todas as repetições, garantindo que cada palavra apareça só uma vez.
- sort(...): Organiza as palavras resultantes em ordem alfabética.
- vocab <-: Guarda esse vetor ordenado e sem duplicatas na variável vocab.

sort(freq, decreasing = TRUE)[1:6]

- sort(freq, decreasing = TRUE): Ordena os valores da variável freq em ordem decrescente (do maior para o menor). Sem o decreasing = TRUE, a ordenação padrão seria do menor para o maior.
- [1:6]: Fatora/fatia a resposta para selecionar apenas do 1º ao 6º elemento dessa lista já ordenada.


### 2. Prever e Explorar

- Inverter a ordem de ordenação

**Previsão:** Ao trocar *decreasing = TRUE* por *decreasing = FALSE* mantendo o corte [1:6], o R deixará de mostrar os termos mais comuns e exibirá as 6 palavras menos frequentes (com menor contagem) do corpus.

In [ ]:
sort(freq, decreasing = FALSE)[1:6]


    acelera aprendizado  artificial   avaliacao        bm25    capturam 
          1           1           1           1           1           1 

**Resultado:** A saída exibe termos raras ou únicos no texto (frequência igual a 1), organizados em ordem crescente de contagem.

## Bloco 4: Matriz Termo-Documento

In [ ]:
tdm <- sapply(tokens, function(tk) {
  as.integer(table(factor(tk, levels = vocab)))
})
rownames(tdm) <- vocab
tdm[1:6, ]

,d1,d2,d3,d4,d5,d6,d7,d8,d9
a,0,0,0,1,1,1,2,0,0
acelera,0,0,0,0,1,0,0,0,0
aprendizado,0,0,0,1,0,0,0,0,0
artificial,0,0,0,0,0,0,0,0,1
avaliacao,0,0,0,0,0,0,1,0,0
bm25,0,0,1,0,0,0,0,0,0


### 1. O que cada linha faz:
- tdm <- sapply(tokens, function(tk) { ... }): aplica uma função a cada elemento de tokens e combina os resultados em uma matriz (Term-Document Matrix).
- table(factor(tk, levels = vocab)): garante que a contagem de frequências considere todo o vocabulário definido, mesmo os termos ausentes no documento específico (que receberão contagem 0).
- as.integer(...): converte a tabela de frequências para um vetor numérico de inteiros puramente limpo.
- rownames(tdm) <- vocab: nomeia as linhas da matriz tdm com as palavras correspondentes contidas em vocab.
- tdm[1:6, ]: exibe um recorte das 6 primeiras linhas (termos) da matriz para todos os documentos.

### 2. Prever e Explorar

- Alterar o recorte de exibição da matriz para os últimos 5 termos do vocabulário

**Previsão:** Ao trocar a indexação [1:6, ] por [(nrow(tdm)-4):nrow(tdm), ], a consulta exibirá as últimas 5 linhas (palavras no fim do alfabeto contidas no vocabulário) e a frequência delas em cada documento.

In [ ]:
tdm[(nrow(tdm) - 4):nrow(tdm), ]

,d1,d2,d3,d4,d5,d6,d7,d8,d9
semantica,0,0,0,0,0,1,0,0,0
texto,0,0,1,0,0,0,0,0,0
um,0,0,1,0,0,0,0,0,0
vetores,0,1,0,0,0,0,0,0,0
vetorial,0,1,0,0,0,0,0,0,0


**Resultado:** Exibição detalhada da matriz focada apenas nas últimas 5 palavras alfabéticas do vocabulário construído ao longo dos documentos.

## Bloco 5: Busca Booleana Simples

In [ ]:
busca_booleana <- function(termo, tdm) {
  termo <- tolower(termo)
  if (!termo %in% rownames(tdm)) return(character(0))
  colnames(tdm)[tdm[termo, ] > 0]
}

busca_booleana("documentos", tdm)
busca_booleana("busca", tdm)

[1] "d1" "d2" "d5" "d6"

[1] "d5" "d7"

### 1. O que cada linha faz:
- busca_booleana <- function(termo, tdm): cria uma função de busca exata que aceita um termo de pesquisa e a matriz termo-documento.
- termo <- tolower(termo): padroniza a entrada da busca convertendo o termo para minúsculas.
- if (!termo %in% rownames(tdm)) return(character(0)): checa se o termo está presente no vocabulário (nomes de linha da tdm). Caso não esteja, retorna um vetor de caracteres vazio de imediato.
- colnames(tdm)[tdm[termo, ] > 0]: verifica na linha do termo onde o valor de frequência é maior que 0 e retorna os nomes das colunas (IDs dos documentos) que satisfazem a condição.

### 2. Prever e Explorar

- Testar uma busca por um termo ausente no corpus (ex: "python") e outro presente com maiúsculas (ex: "CIENCIA")

**Previsão:** A busca por "python" retornará character(0) por não constar no vocabulário, enquanto a busca por "CIENCIA" deve tratar o texto para minúsculas e encontrar com sucesso os documentos associados.

In [ ]:
print(busca_booleana("python", tdm))
print(busca_booleana("CIENCIA", tdm))

character(0)
[1] "d8" "d9"


**Resultado:** A consulta para "python" respondeu character(0) por falta de correspondência exata, e "CIENCIA" converteu com sucesso identificando os documentos corretos.

## Bloco 6: TF-IDF

In [ ]:
N <- ncol(tdm)
df <- rowSums(tdm > 0)   # em quantos docs cada termo aparece
idf <- log(N / df)       # inverse document frequency
tfidf <- tdm * idf       # recicla idf por linha

round(tfidf[c("documentos", "recuperacao", "busca", "de"), ], 2)

,d1,d2,d3,d4,d5,d6,d7,d8,d9
documentos,0.81,0.81,0.00,0.0,0.81,0.81,0.0,0.00,0.00
recuperacao,1.50,0.00,0.00,1.5,0.00,0.00,0.0,0.00,0.00
busca,0.00,0.00,0.00,0.0,1.50,0.00,1.5,0.00,0.00
de,0.41,0.41,0.81,0.0,0.00,0.41,0.0,0.41,0.41


### 1. O que cada linha faz:
- N <- ncol(tdm): obtém o número total de documentos analisados contanto o número de colunas da matriz.
- df <- rowSums(tdm > 0): calcula o *Document Frequency* contando em quantos documentos cada termo aparece ao menos uma vez (valor > 0).
- idf <- log(N / df): calcula a *Inverse Document Frequency* aplicando o logaritmo natural do total de documentos dividido pela frequência em documentos. Palavras raras ganham peso maior, palavras comuns ganham peso menor.
- tfidf <- tdm * idf: calcula a matriz TF-IDF ponderando a frequência do termo pela sua importância global no corpus.
- round(tfidf[...], 2): recorta os termos específicos indicados e arredonda seus pesos TF-IDF para 2 casas decimais para facilitar a leitura.

### 2. Prever e Explorar

- Modificar a base do logaritmo para base 10 (`log10`) no cálculo da IDF

**Previsão:** Ao usar `log10(N / df)` no lugar do logaritmo natural (`log`), a escala de relevância dos pesos reduzirá proporcionalmente no valor numérico absoluto, mas manterá a exata mesma relação proporcional relativa entre a raridade dos termos.

In [ ]:
idf_log10 <- log10(N / df)
tfidf_log10 <- tdm * idf_log10
round(tfidf_log10[c("documentos", "recuperacao", "busca", "de"), ], 2)

,d1,d2,d3,d4,d5,d6,d7,d8,d9
documentos,0.35,0.35,0.00,0.00,0.35,0.35,0.00,0.00,0.00
recuperacao,0.65,0.00,0.00,0.65,0.00,0.00,0.00,0.00,0.00
busca,0.00,0.00,0.00,0.00,0.65,0.00,0.65,0.00,0.00
de,0.18,0.18,0.35,0.00,0.00,0.18,0.00,0.18,0.18


**Resultado:** Os valores de peso TF-IDF reduziram em amplitude (por conta da mudança para base 10), porém a penalização da palavra hipercomum "de" se manteve idêntica em zero (já que log10(N/N) = 0).

## Bloco 7: Baixar um corpus real da Wikipédia

In [ ]:
library(httr2)

baixar_wiki <- function(titulo) {
  request("https://pt.wikipedia.org/w/api.php") |>
    req_url_query(action = "query", prop = "extracts", explaintext = 1,
                  format = "json", redirects = 1, titles = titulo) |>
    req_perform() |> resp_body_json() |>
    (\(r) r$query$pages[[1]]$extract)()
}

texto <- baixar_wiki("Santos (São Paulo)")
substr(texto, 1, 60)

[1] "Santos é um município brasileiro no litoral do estado de São"

### 1. O que cada linha faz:
- library(httr2): carrega a biblioteca `httr2` encarregada de fazer requisições HTTP RESTful de forma moderna no R.
- baixar_wiki <- function(titulo) { ... }: define uma função para realizar requisições diretas à API oficial da Wikipédia.
- request("...") |> req_url_query(...): monta o endpoint base da API e configura parâmetros da query (extrair texto limpo, formato JSON, resolver redirecionamentos).
- req_perform() |> resp_body_json(): executa a chamada à API web e decodifica a resposta diretamente como uma estrutura de lista JSON.
- (\(r) r$query$pages[[1]]$extract)(): função anônima que navega na árvore do JSON retornado e isola somente o texto extraído da página.
- substr(texto, 1, 60): recorta e exibe apenas os primeiros 60 caracteres da string retornada.

### 2. Prever e Explorar

- Alterar a busca para um novo tópico da Wikipédia e aumentar o tamanho do recorte exibido para 120 caracteres

**Previsão:** O código fará uma nova requisição à Wikipédia buscando pelo verbete informado e retornará o trecho introdutório do artigo com o dobro de tamanho de caracteres.

In [ ]:
texto_novo <- baixar_wiki("Inteligência artificial")
substr(texto_novo, 1, 120)

[1] "Inteligência artificial (frequentemente abreviada por IA ou AI, do inglês artificial intelligence) é, de forma genérica,"

**Resultado:** O texto foi baixado em tempo real via API com sucesso e o recorte exibiu perfeitamente os primeiros 120 caracteres da introdução do artigo sobre Inteligência Artificial.

## Bloco 8: Regex vs. Índice de termos

In [ ]:
grep("recupera", docs)             # acha "recuperacao": casa SUBSTRING
busca_booleana("recupera", tdm)    # nao acha: o TERMO nao existe

[1] 1 4

character(0)

### 1. O que cada linha faz:
- grep("recupera", docs): realiza uma busca textual baseada em Expressão Regular (Regex) procurando o trecho exato como substring em qualquer parte das frases do vetor `docs`. Retorna os índices numéricos dos documentos onde o trecho aparece.
- busca_booleana("recupera", tdm): tenta realizar a busca booleana por tokenização/vocabulário no índice exato da matriz `tdm`. Falha por não encontrar o termo exato "recupera", visto que no vocabulário a palavra tokenizada consta apenas como "recuperacao".

### 2. Prever e Explorar

- Testar uma busca parcial por "estatist" usando `grep` e `busca_booleana`

**Previsão:** O `grep` encontrará com sucesso os documentos correspondentes por casar a substring "estatist" presente em "estatistica" / "estatistico", enquanto a `busca_booleana` retornará `character(0)` devido ao radical incompleto não ser um token catalogado no vocabulário.

In [ ]:
grep("estatist", docs)
busca_booleana("estatist", tdm)

[1] 4 8

character(0)

**Resultado:** `grep` retornou os documentos contendo as derivações da palavra por casamento de padrão de texto, ao passo que `busca_booleana` confirmou a rigidez da busca por vocabulário/tokens inteiros.

# Parte 2: *corpus* da Baixada Santista

1. Baixe da Wikipédia os artigos de 3 municípios da região (comece pequeno).
2. Guarde-os num vetor nomeado docs, como fizemos em aula.
3. Tokenize, monte o vocabulário e responda: quantos termos distintos surgiram?
Compare com os 45 do nosso corpus de brinquedo.
4. Liste os 10 termos mais frequentes. Eles são informativos? O que isso sugere sobre a próxima aula?

Cite a fonte (CC BY-SA). Esse corpus será a base do seu motor no restante do semestre.

---

Nesta etapa, aplicamos os conceitos desenvolvidos na Parte 1 a um corpus real extraído da Wikipédia em Português. As cidades selecionadas para a análise comparativa são **Santos**, **Cubatão** e **Guarujá**.

### Objetivos:
1. Baixar o texto dos verbetes das três cidades via API da Wikipédia.
2. Fazer o pré-processamento (limpeza, remoção de pontuação, tokenização e remoção de *stopwords*).
3. Construir o vocabulário e a Matriz Termo-Documento (TDM).
4. Calcular a matriz TF-IDF para ponderar a importância dos termos por cidade.
5. Executar buscas booleanas e identificar as palavras com maior peso de relevância em cada município.

## Bloco 1: Coleta dos Verbetes das Cidades

In [ ]:
library(httr2)

baixar_wiki <- function(titulo) {
  request("https://pt.wikipedia.org/w/api.php") |>
    req_url_query(action = "query", prop = "extracts", explaintext = 1,
                  format = "json", redirects = 1, titles = titulo) |>
    req_perform() |> resp_body_json() |>
    (\(r) r$query$pages[[1]]$extract)()
}

# Coleta dos textos para Santos, Cubatão e Guarujá
cidades_docs <- c(
  Santos  = baixar_wiki("Santos"),
  Cubatao = baixar_wiki("Cubatão"),
  Guaruja = baixar_wiki("Guarujá")
)

sapply(cidades_docs, nchar)

Santos Cubatao Guaruja 
  54140   16597   19274

## Bloco 2: Limpeza, Tokenização e Remoção de Stopwords

In [ ]:
# Lista expandida de stopwords em português
# (Lista adaptada a partir da lista de stopwords do Snowball/NLTK)
stopwords_pt <- c(
  "a", "ao", "aos", "aquela", "aquelas", "aquele", "aqueles", "aquilo", "as", "ate", "com", "como", "da", "das",
  "de", "dela", "delas", "dele", "deles", "depois", "do", "dos", "e", "e", "ela", "elas", "ele", "eles", "em",
  "entre", "era", "eram", "eramos", "essa", "essas", "esse", "esses", "esta", "estamos", "estao", "estar", "estas",
  "estava", "estavam", "estavamos", "este", "esteja", "estejam", "estejamos", "estes", "esteve", "estive", "estivemos",
  "estiver", "estiveram", "estiveremos", "estiverem", "estivermos", "estivesse", "estivessem", "estivessemos", "estou",
  "eu", "foi", "fomos", "for", "fora", "foram", "forem", "formos", "fosse", "fossem", "fossemos", "ha", "haja",
  "hajam", "hajamos", "havemos", "haver", "haveria", "haveriam", "haveriamos", "haveremos", "haverao", "hei", "houve",
  "houver", "houvera", "houveram", "houverei", "houverem", "houveremos", "houveria", "houveriam", "houveriamos",
  "houvermos", "houvesse", "houvessem", "houvessemos", "isso", "isto", "já", "ja", "lhe", "lhes", "mais", "mas",
  "me", "mesmo", "meu", "meus", "minha", "minhas", "muito", "na", "nao", "nas", "nem", "no", "nos", "nossa",
  "nossas", "nosso", "nossos", "num", "numa", "o", "os", "ou", "para", "pela", "pelas", "pelo", "pelos", "por",
  "qual", "quando", "que", "quem", "se", "seja", "sejam", "sejamos", "sem", "sendo", "ser", "sera", "serao",
  "serei", "seremos", "seria", "seriam", "seriamos", "seu", "seus", "so", "somos", "sou", "sua", "suas", "tambem",
  "te", "tem", "temos", "tenho", "tenha", "tenham", "tenhamos", "tenha", "ter", "tera", "terao", "terei", "teremos",
  "teria", "teriam", "teriamos", "teve", "tinha", "tinham", "tinhamos", "tive", "tiver", "tiveram", "tiveremos",
  "tiverem", "tivermos", "tivesse", "tivessem", "tivessemos", "tu", "tua", "tuas", "um", "uma", "você", "vocês",
  "vos", "sao", "sobre", "segundo", "onde", "partir", "alem", "possui", "foram", "após", "apos"
)

tokenizar_limpo <- function(texto) {
  # Passa para minúsculas
  texto <- tolower(texto)
  # Remove acentuação/diacríticos (sinais que mudam o som) usando iconv
  texto <- iconv(texto, to = "ASCII//TRANSLIT")
  # Remove pontuações, números e caracteres especiais deixando apenas letras e espaços
  texto <- gsub("[^a-z\\s]", " ", texto)
  # Quebra em tokens por espaços em branco
  tk <- unlist(strsplit(texto, "\\s+"))
  # Remove tokens vazios, de apenas 1 letra e stopwords
  tk <- tk[nchar(tk) > 1]
  tk <- tk[!tk %in% stopwords_pt]
  return(tk)
}

tokens_cidades <- lapply(cidades_docs, tokenizar_limpo)
sapply(tokens_cidades, length)

Santos Cubatao Guaruja 
   4881    1478    1728

## Bloco 3: Construção da Matriz Termo-Documento (TDM)

In [ ]:
# Vocabulário único ordenado de todo o corpus das três cidades
vocab_cidades <- sort(unique(unlist(tokens_cidades)))
cat("Tamanho do vocabulário único:", length(vocab_cidades), "\n")

# Criação da TDM
tdm_cidades <- sapply(tokens_cidades, function(tk) {
  as.integer(table(factor(tk, levels = vocab_cidades)))
})
rownames(tdm_cidades) <- vocab_cidades

# Seleção manual, baseada no conhecimento prévio sobre as características socioeconômicas de cada cidade
termos_amostra <- c("porto", "industria", "praia", "turismo", "petroleo", "ferrovia")
termos_validos <- termos_amostra[termos_amostra %in% rownames(tdm_cidades)]
tdm_cidades[termos_validos, ]

Tamanho do vocabulário único: 3428 


,Santos,Cubatao,Guaruja
porto,28,4,5
industria,1,0,2
praia,18,0,9
turismo,9,0,9
petroleo,1,0,0
ferrovia,1,0,0


## Bloco 4: Cálculo da Matriz TF-IDF

In [ ]:
N_cid <- ncol(tdm_cidades)
df_cid <- rowSums(tdm_cidades > 0)
idf_cid <- log(N_cid / df_cid)
tfidf_cidades <- tdm_cidades * idf_cid

# Exibição do TF-IDF para a mesma amostra de termos
round(tfidf_cidades[termos_validos, ], 2)

,Santos,Cubatao,Guaruja
porto,0.00,0,0.00
industria,0.41,0,0.81
praia,7.30,0,3.65
turismo,3.65,0,3.65
petroleo,1.10,0,0.00
ferrovia,1.10,0,0.00


## Bloco 5: Extração dos Termos Mais Relevantes de Cada Cidade

Ao usarmos o TF-IDF, os termos que aparecem em todas as três cidades recebem penalização (IDF menor), enquanto os termos que caracterizam a identidade específica de cada cidade ganham destaque numérico.

In [ ]:
top_termos_cidade <- function(cidade, top_n = 10) {
  pesos <- tfidf_cidades[, cidade]
  head(sort(pesos, decreasing = TRUE), top_n)
}

cat("--- TOP TERMOS: SANTOS ---\n")
print(round(top_termos_cidade("Santos", 8), 2))

cat("\n--- TOP TERMOS: CUBATÃO ---\n")
print(round(top_termos_cidade("Cubatao", 8), 2))

cat("\n--- TOP TERMOS: GUARUJÁ ---\n")
print(round(top_termos_cidade("Guaruja", 8), 2))

--- TOP TERMOS: SANTOS ---
  teatro     cafe     zona  pessoas   serrat festival    nomes    maior 
   26.37    12.08    12.08    10.99    10.99     9.89     9.89     9.33 

--- TOP TERMOS: CUBATÃO ---
      sala       vice passarelli   sesmaria   cruzeiro    fazenda     marcia 
      6.59       6.59       5.49       5.49       4.39       4.39       4.39 
  vereador 
      4.39 

--- TOP TERMOS: GUARUJÁ ---
      carece      guaruja        amaro       fontes        hotel pitangueiras 
       17.58        13.38        12.08         6.89         6.59         6.59 
       canal      enseada 
        4.39         4.39 


**Observação:** Alguns dos termos com maior peso TF-IDF para Cubatão não são exatamente palavras temáticas, mas sim nomes próprios (ex: "passarelli", "marcia", "vereador", "sesmaria"), provavelmente citados no verbete em trechos sobre história ou política local. Isso acontece porque o TF-IDF favorece termos raros que aparecem em poucos documentos, e um nome próprio mencionado só no artigo de Cubatão recebe um IDF alto, mesmo sem representar uma característica geral da cidade.

Esse é um limite conhecido do TF-IDF puro: ele não distingue "termo temático relevante" de "entidade nomeada rara", o que exigiria uma etapa adicional de reconhecimento de entidades nomeadas (NER) para ser filtrado.

## Bloco 6: Consultas e Busca Booleana

In [ ]:
busca_cidades <- function(termo, tdm) {
  termo <- tolower(termo)
  termo <- iconv(termo, to = "ASCII//TRANSLIT")
  if (!termo %in% rownames(tdm)) return("Termo não encontrado no vocabulário")

  docs_encontrados <- colnames(tdm)[tdm[termo, ] > 0]
  return(docs_encontrados)
}

# Testando buscas booleanas no corpus real
cat("Busca por 'porto':", busca_cidades("porto", tdm_cidades), "\n")
cat("Busca por 'industria':", busca_cidades("industria", tdm_cidades), "\n")
cat("Busca por 'ecoturismo':", busca_cidades("ecoturismo", tdm_cidades), "\n")

Busca por 'porto': Santos Cubatao Guaruja 
Busca por 'industria': Santos Guaruja 
Busca por 'ecoturismo': Guaruja 


# Atualização (25/08/2026): Adição de Similaridade de Cosseno (Aulas 1.5 e 2)

Aplicação da similaridade do cosseno, dando continuidade à pesquisa iniciada na Aula 01 com os verbetes da Wikipedia de Santos, Cubatão e Guarujá, nos mesmos moldes de um trabalho já feito em outra disciplina (o projeto "Similaridade do Cosseno" da Flávia, que usou TF-IDF + similaridade do cosseno para ranquear filmes por proximidade com uma query de preferência do usuário. A lógica é a mesma; aqui o "documento" é o verbete de uma cidade, e a "query" é um interesse de busca (ex: turismo, indústria, porto).

Fonte dos textos: Wikipedia em português (CC BY-SA), mesmos verbetes coletados na Aula 01.

## Bloco 1: Recuperando o corpus das 3 cidades

In [1]:
library(httr2)

baixar_wiki <- function(titulo) {
  request("https://pt.wikipedia.org/w/api.php") |>
    req_url_query(action = "query", prop = "extracts", explaintext = 1,
                  format = "json", redirects = 1, titles = titulo) |>
    req_perform() |> resp_body_json() |>
    (\(r) r$query$pages[[1]]$extract)()
}

cidades_docs <- c(
  Santos  = baixar_wiki("Santos"),
  Cubatao = baixar_wiki("Cubatao"),
  Guaruja = baixar_wiki("Guaruja")
)
sapply(cidades_docs, nchar)

Santos Cubatao Guaruja 
  54140   16597   19274

Na Aula 01 os tamanhos obtidos foram Santos=54140, Cubatão=16597 e Guarujá=19274 caracteres. Os valores podem variar um pouco se o verbete tiver sido editado desde então.

## Bloco 2: Limpeza, tokenização e TDM (reaproveitando a função da Aula 01)

In [2]:
stopwords_pt <- c(
  "a","ao","aos","aquela","aquelas","aquele","aqueles","aquilo","as","ate","com","como","da","das",
  "de","dela","delas","dele","deles","depois","do","dos","e","ela","elas","ele","eles","em",
  "entre","era","eram","essa","essas","esse","esses","esta","estamos","estao","estar","estas",
  "estava","estavam","este","estes","eu","foi","fomos","for","fora","foram","ha","isso","isto",
  "ja","lhe","lhes","mais","mas","me","mesmo","meu","meus","minha","minhas","muito","na","nao",
  "nas","nem","no","nos","nossa","nossas","nosso","nossos","num","numa","o","os","ou","para",
  "pela","pelas","pelo","pelos","por","qual","quando","que","quem","se","seja","sejam","sem",
  "sendo","ser","sera","seu","seus","so","somos","sou","sua","suas","tambem","te","tem","temos",
  "tenho","ter","tera","teve","tinha","tinham","tu","tua","tuas","um","uma","voce","voces",
  "vos","sao","sobre","segundo","onde","partir","alem","possui","apos"
)

tokenizar_limpo <- function(texto) {
  texto <- tolower(texto)
  texto <- iconv(texto, to = "ASCII//TRANSLIT")
  texto <- gsub("[^a-z\\s]", " ", texto)
  tk <- unlist(strsplit(texto, "\\s+"))
  tk <- tk[nchar(tk) > 1]
  tk <- tk[!tk %in% stopwords_pt]
  tk
}

tokens_cidades <- lapply(cidades_docs, tokenizar_limpo)
vocab_cidades  <- sort(unique(unlist(tokens_cidades)))

tdm_cidades <- sapply(tokens_cidades, function(tk) {
  as.integer(table(factor(tk, levels = vocab_cidades)))
})
rownames(tdm_cidades) <- vocab_cidades

N_cid  <- ncol(tdm_cidades)
df_cid <- rowSums(tdm_cidades > 0)
idf_cid <- log(N_cid / df_cid)
tfidf_cidades <- tdm_cidades * idf_cid

dim(tfidf_cidades)

[1] 3438    3

**Nota:**  Reaproveita a mesma lista de stopwords e a mesma função de limpeza (minúsculas, remoção de acentos, remoção de pontuação) já validada na Aula 01.

## Bloco 3: Busca por similaridade do cosseno no corpus real

In [3]:
cosseno <- function(a, b) {
  denom <- sqrt(sum(a^2)) * sqrt(sum(b^2))
  if (denom == 0) return(0)   # documento sem nenhum termo da consulta -> similaridade 0, não NaN
  sum(a * b) / denom
}

buscar_cidade <- function(consulta_txt, tfidf, vocab, idf) {
  termos_q <- tokenizar_limpo(consulta_txt)
  q  <- as.integer(table(factor(termos_q, levels = vocab)))
  qw <- q * idf
  scores <- apply(tfidf, 2, function(v) cosseno(qw, v))
  sort(scores, decreasing = TRUE)
}

cat("=== Consulta: porto industria petroleo ===\n")
print(round(buscar_cidade("porto industria petroleo", tfidf_cidades, vocab_cidades, idf_cid), 4))

cat("\n=== Consulta: praia turismo ===\n")
print(round(buscar_cidade("praia turismo", tfidf_cidades, vocab_cidades, idf_cid), 4))

cat("\n=== Consulta: poluicao industrial saude ===\n")
print(round(buscar_cidade("poluicao industrial saude", tfidf_cidades, vocab_cidades, idf_cid), 4))

=== Consulta: porto industria petroleo ===
 Santos Guaruja Cubatao 
 0.0134  0.0064  0.0000 

=== Consulta: praia turismo ===
Guaruja  Santos Cubatao 
 0.1181  0.0883  0.0000 

=== Consulta: poluicao industrial saude ===
Guaruja Cubatao  Santos 
 0.0197  0.0158  0.0033 


**Nota:** A função buscar_cidade é exatamente a mesma lógica do sistema de recomendação da Flávia: query -> TF-IDF -> cosseno -> ranking; só que sem o filtro de idioma/nota que ela usou para filmes, e com uma consulta curta em vez de uma descrição livre.

## Bloco 4: Verificação com dados já conhecidos da Aula 01 (Execução offline)

In [4]:
# Sub-vetores reais já calculados na Aula 01 (Bloco 4: Cálculo da Matriz TF-IDF),
# usados aqui só para demonstrar o método sem depender de internet em outro ambiente.
termos_amostra <- c("porto","industria","praia","turismo","petroleo","ferrovia")
tfidf_amostra <- cbind(
  Santos  = c(0.00, 0.41, 7.30, 3.65, 1.10, 1.10),
  Cubatao = c(0,    0,    0,    0,    0,    0   ),
  Guaruja = c(0.00, 0.81, 3.65, 3.65, 0.00, 0.00)
)
rownames(tfidf_amostra) <- termos_amostra

idf_amostra <- c(porto = 0, industria = log(3/2), praia = log(3/2),
                  turismo = log(3/2), petroleo = log(3/1), ferrovia = log(3/1))

buscar_amostra <- function(termos_consulta) {
  q  <- as.integer(rownames(tfidf_amostra) %in% termos_consulta)
  qw <- q * idf_amostra[rownames(tfidf_amostra)]
  scores <- apply(tfidf_amostra, 2, function(v) cosseno(qw, v))
  sort(scores, decreasing = TRUE)
}

cat("=== Consulta: porto + industria + petroleo ===\n")
print(round(buscar_amostra(c("porto","industria","petroleo")), 4))

cat("\n=== Consulta: praia + turismo ===\n")
print(round(buscar_amostra(c("praia","turismo")), 4))

=== Consulta: porto + industria + petroleo ===
 Santos Guaruja Cubatao 
 0.1411  0.0537  0.0000 

=== Consulta: praia + turismo ===
Guaruja  Santos Cubatao 
 0.9879  0.9308  0.0000 


**Leitura:**
- Na consulta **"porto industria petroleo"**, Santos lidera (0.1411), sendo o único dos três verbetes que menciona "petróleo"/"ferrovia" e tem o maior peso relativo em "indústria". Guarujá aparece bem atrás (0.0537) e Cubatão empata em zero.
- Na consulta **"praia turismo"**, a ordem se inverte: Guarujá assume a frente (0.9879), com Santos logo atrás (0.9308). Os dois têm quase a mesma composição relativa de "praia"+"turismo", então o ângulo entre seus vetores e o da consulta é quase idêntico.
- **Cubatão fica em zero nas duas consultas.** E aqui a Aula 1,5 se conecta direto com a Aula 02: nesse recorte de 6 termos, o vetor de Cubatão é literalmente o vetor nulo. O verbete não usa nenhuma dessas 6 palavras específicas (mesmo sendo historicamente a cidade mais industrial das três). Cosseno com o vetor zero não tem direção definida, então a função retorna 0 por convenção. Isso não significa que Cubatão seja "irrelevante" para indústria. Significa que, neste recorte manual de 6 termos, o modelo não tem vocabulário para enxergar a cidade. É o mesmo problema do "motor cego" das rotas (Aula 1,5, Bloco 9): sem contraste de vocabulário, não há distância para medir. Rodando a busca sobre o vocabulário completo (~3428 termos, Bloco 3 acima) esse problema deve diminuir bastante, porque Cubatão certamente tem termos próprios de peso alto (ex: os nomes citados na Aula 01, "sesmaria", "vereador", ainda que esses tragam o problema oposto, de nomes próprios com IDF artificialmente alto).

---
## Conclusão

- A busca booleana da Aula 01 só dizia **se** um termo aparecia; a similaridade do cosseno da Aula 02 diz **o quanto** um documento combina com a consulta, permitindo ranquear.
- O mesmo padrão TF-IDF -> vetor de consulta -> cosseno -> ranking que a Flávia usou para recomendar filmes funciona aqui para recuperar o verbete de cidade mais alinhado com um interesse de busca.
- O limite fica evidente quando o vocabulário é pequeno ou pouco representativo (o caso de Cubatão no recorte de 6 termos): sem contraste, o cosseno não discrimina, o mesmo limite mostrado com as rotas na Aula 1,5.
- Próximo passo natural (Aula 03 em diante): pré-processamento mais cuidadoso do corpus real e, mais adiante, BM25/embeddings para reduzir esse tipo de ponto cego.
